soc/xtensa/intel_adsp/common/ipc.c

```diff
@@ -205,12 +218,157 @@ static int dt_init(const struct device *dev)
+
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->resume_fn = fn;
+	api->resume_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg)
+{
+	struct ipc_control_driver_api *api =
+		(struct ipc_control_driver_api *)dev->api;
+	struct intel_adsp_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	api->suspend_fn = fn;
+	api->suspend_fn_args = arg;
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+/**
+ * @brief Waits for IPC device to become non-busy.
+ * Peforms @see retries number of checks for the busy status each time after @see usec_to_wait
+ * elapsed.
+ *
+ * @param usec_to_wait Time to wait in microseconds before checking again for the busy status.
+ * @param retries Number of retries Number of times to repeat a busy check.
+ * @return true Returned when device became non-busy during expected timeline.
+ * @return false Returned when device is still busy when expected timeline elapsed.
+ */
+static bool wait_ipc_device_not_busy(uint32_t usec_to_wait, uint32_t retries)
+{
+	if (pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV)) {
+		while(retries > 0) {
+			k_busy_wait(usec_to_wait);
+			if (!pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV))
+				return true;
+			retries--;
+		}
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * @brief Manages IPC driver power state change.
+ *
+ * @param dev IPC device.
+ * @param action Power state to be changed to.
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks. Returning error code from functions
+ * invoked during power stansition, does not release the pm_device_state_lock.
+ * It must be released manually by the caller during error handling process.
+ *
+ * @note Power state of the IPC Device is changed atomically.
+ * PM lock is taken at the start of each power transition to prevent power state change
+ * if the process fails.
+ * If IPC Device performs hardware operation, this function waits for 5 seconds in total
+ * for IPC Device to become non-busy (to finish all pending transactions).
+ * Note: when devices change power state using pm_suspend_devices() function, busy devices are
+ * ignored. However, calling pm_device_action_run() directly on the device, does not prevent
+ * busy device from power transition.
+ */
+static int ipc_pm_action(const struct device *dev, enum pm_device_action action)
+{
+	const struct ipc_control_driver_api *api =
+		(const struct ipc_control_driver_api *)dev->api;
+
+	/* wait up to 5 seconds for IPC device to become non-busy */
+	if(!wait_ipc_device_not_busy(1000000, 5))
+		return -EBUSY;
```

Comment:
I see this logic here, do you need `intel_adsp_ipc_resume_handler_t`  and `intel_adsp_ipc_suspend_handler_t` ? My understand that that API is because the IPC message can vary depending on target, but isn't this block the same ?

Can't we make simple it and having the msg logic in this driver instead of that callback ? That is a pattern that is not common in our drivers, maybe the right thing to do is having a kernel event for when the device change the status and the application has a listener for it where it does whathever it wants instead of doing it inside the pm action.

Reply:
@ceolin a right thing to do is to move IPC driver to Zephyr completely. However, this is not possible now until IPC3 exists (that does not use this Zephyr driver). Callbacks you mentioned are defined just because we can not #include sof headers in the Zephyr code and some key IPC elements are in the SOF.
So we do this "the hard way":
* define callback register function in Zephyr
* call callback register function in SOF and assign some SOF IPC function to it.
* Zephyr driver in some point calls callback.
* thanks to this, steering returns to SOF (some IPC function)
* then we do what we need to do (and we reference elements that are defined on SOF side eg. ipc message queue, ipc flags, ipc tasks)

Take a look how other callbacks are registered in this API (eg. isr). Adding two more is nothing new really.

Reply:
That is what I thought, understand the point and I am fine with that.